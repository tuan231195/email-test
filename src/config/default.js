require('dotenv').config();

module.exports = {
	logging: {
		logLevel: 'debug',
	},
	email: {
		mailgun: {
			apiToken: process.env.MAILGUN_API_KEY,
			domain: process.env.MAILGUN_DOMAIN,
		},
		sendgrid: {
			apiToken: process.env.SENDGRID_API_KEY,
		},
		errorThreshold: 5,
		queueUrl:
			'https://sqs.ap-southeast-2.amazonaws.com/923678104243/email-queue.fifo',
	},
	db: {
		redis: {
			host: 'localhost',
			port: 6379,
			password: '',
		},
	},
};
