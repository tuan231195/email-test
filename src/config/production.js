const { asyncConfig } = require('./utils');
const environment = 'production';

module.exports = {
	logging: {
		logLevel: 'info',
	},
	email: {
		mailgun: {
			apiToken: asyncConfig(environment, 'email.mailgun.apiToken'),
			domain: asyncConfig(environment, 'email.mailgun.domain'),
		},
		sendgrid: {
			apiToken: asyncConfig(environment, 'email.sendgrid.apiToken'),
		},
		queueUrl:
			'https://sqs.ap-southeast-2.amazonaws.com/923678104243/email-queue.fifo',
	},
	db: {
		redis: {
			password: asyncConfig(environment, 'db.redis.password'),
			host: asyncConfig(environment, 'db.redis.host'),
			port: 16551,
		},
	},
};
