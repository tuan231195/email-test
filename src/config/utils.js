const AWS = require('aws-sdk');
const { get } = require('lodash');
const { asyncConfig: loadAsyncConfig } = require('config/async');

const secretsCache = {};

const secretsManager = new AWS.SecretsManager({
	region: 'ap-southeast-2',
});

exports.loadSecret = function(environment) {
	if (secretsCache[environment]) {
		return secretsCache[environment];
	}
	secretsCache[environment] = secretsManager
		.getSecretValue({
			SecretId: `secrets-${environment}`,
		})
		.promise()
		.then(({ SecretString: secret }) => JSON.parse(secret));
	return secretsCache[environment];
};

exports.getSecretKey = async function(environment, key) {
	const secret = await exports.loadSecret(environment);
	return get(secret, key);
};

exports.asyncConfig = function(environment, key) {
	return loadAsyncConfig(exports.getSecretKey(environment, key));
};
