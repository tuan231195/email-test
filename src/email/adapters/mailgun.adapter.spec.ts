import { MailgunAdapter } from './mailgun.adapter';
import { mockConfig } from '../../test/config';
import { mockAxios } from '../../test/axios';
import { mockLogger } from '../../test/logger';
import testCase from './test/test.case';

jest.mock('config');

describe('Testing mailgun adapter', () => {
	let adapter: MailgunAdapter, axios;
	beforeEach(() => {
		axios = mockAxios();
		const httpClient: any = {
			instance() {
				return axios;
			},
		};

		const config = mockConfig({
			email: {
				mailgun: {
					domain: 'test',
					apiToken: 'token',
				},
			},
		});
		const logger = mockLogger();
		adapter = new MailgunAdapter(config, httpClient, logger);
	});

	test('Should call http client post with the right data', async () => {
		await adapter.sendEmail(testCase);
		expect(axios.postForm).toHaveBeenCalled();
		expect(axios.postForm.mock.calls).toMatchSnapshot();
	});
});
