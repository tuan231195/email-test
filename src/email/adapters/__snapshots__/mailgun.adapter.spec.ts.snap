// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing mailgun adapter Should call http client post with the right data 1`] = `
Array [
  Array [
    "https://api.mailgun.net/v3/test/messages",
    Object {
      "bcc": "Nguyen <vdtn359@gmail.com>",
      "cc": "TN <vdtn359@uowmail.edu.au>",
      "from": "Tuan Nguyen <vdtn359@gmail.com>",
      "html": "Test body",
      "subject": "Test subject",
      "to": "Tuan <vdtn359@uowmail.edu.au>",
    },
    Object {
      "auth": Object {
        "password": "token",
        "username": "api",
      },
    },
  ],
]
`;
